gcs/src/main/scala/org/broadinstitute/monster/storage/gcs/GcsApi.scala

```diff
@@ -290,44 +290,80 @@ class GcsApi private[gcs] (runHttp: Request[IO] => Resource[IO, Response[IO]]) {
     chunkData: Stream[IO, Byte],
     bytesPerUploadRequest: Int = MaxBytesPerUploadRequest
   ): IO[Either[Long, Unit]] = {
-    val chunks = chunkData.chunkN(bytesPerUploadRequest)
-
-    chunks.zipWithIndex.evalMap {
-      case (chunk, index) =>
-        val chunkSize = chunk.size
-        val chunkStart = bytesPerUploadRequest * index
-
-        val gcsReq = Request[IO](
-          method = Method.PUT,
-          uri = baseGcsUploadUri(bucket, "resumable")
-            .withQueryParam("upload_id", uploadToken),
-          headers = Headers.of(
-            `Content-Length`.unsafeFromLong(chunkSize.toLong),
-            `Content-Range`(chunkStart, chunkStart + chunkSize - 1)
-          ),
-          body = Stream.chunk(chunk)
-        )
 
-        runHttp(gcsReq).use { response =>
-          if (response.status.code == 308) {
-            val bytesReceived = for {
-              range <- response.headers.get(`Content-Range`)
-              rangeEnd <- range.range.second
-            } yield {
-              rangeEnd
-            }
-            IO.pure(Left(bytesReceived.getOrElse(rangeStart + chunkSize)))
-          } else if (response.status.isSuccess) {
-            IO.pure(Right(()))
-          } else {
-            IO.raiseError(
-              new RuntimeException(
-                s"Failed to upload chunk with upload token: $uploadToken"
-              )
+    // Helper method to incrementally push data from the input stream to GCS.
+    def pushChunks(
+      start: Long,
+      data: Stream[IO, Byte]
+    ): Pull[IO, Nothing, Either[Long, Unit]] =
+      // Pull the first chunk of data off the front of the stream.
+      data.pull
+        .unconsN(bytesPerUploadRequest, allowFewer = true)
+        .flatMap {
+          // If the stream was empty, mark that we haven't made any progress
+          // from the initial byte.
+          case None => Pull.pure(Left(start))
+
+          // If a chunk of data was pulled, upload it to GCS.
+          case Some((chunk, remaining)) =>
+            val chunkSize = chunk.size
+            val chunkEnd = start + chunkSize - 1
+
+            val gcsReq = Request[IO](
+              method = Method.PUT,
+              uri = baseGcsUploadUri(bucket, "resumable")
+                .withQueryParam("upload_id", uploadToken),
+              headers = Headers.of(
+                `Content-Length`.unsafeFromLong(chunkSize.toLong),
+                `Content-Range`(start, chunkEnd)
+              ),
+              body = Stream.chunk(chunk)
             )
-          }
+
+            Pull.eval {
+              runHttp(gcsReq).use { response =>
+                // GCS uses a redirect code to signal that more bytes should be uploaded.
+                if (response.status.code == 308) {
+                  val bytesReceived = for {
+                    range <- response.headers.get(Range)
```

Comment:
Is this the source of truth  or there a way to verify that this range is correct. If not you could end up stuck in a loop

Reply:
This is the source of truth (according to Google). The concern about an infinite loop is still valid though. I'll think about how we could add a max-retries paramater.